### PR TITLE
Feature/graph refactor k paths

### PR DESCRIPTION
**Pull request by hkernbach**
### Scope & Purpose

This PR is Part 1 of the graph refactor feature. It covers currently the K_PATHS algorithm execution on a single server.
More partial PR's will follow and will include more algorithms and cluster features.

- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] Design document: https://github.com/arangodb/documents/blob/7ddae6f13a37b4524f32115d345288ce380f7907/DesignDocuments/02_PLANNING/GraphRefactoring.md

### Testing & Verification

*(Please pick either of the following options)*

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### CODE REVIEW

File: arangod/Aql/KShortestPathsExecutor.cpp [hunk 1]
```diff
@@ -265,7 +317,16 @@ auto KShortestPathsExecutor<FinderType>::getVertexId(InputVertex const& vertex,
       AqlValue const& in = row.getValue(vertex.reg);
       if (in.isObject()) {
         try {
-          auto idString = _finder.options().trx()->extractIdString(in.slice());
+          std::string idString;
+          // TODO:  calculate expression once e.g. header constexpr bool and check then here
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsExecutor.cpp:321]: Needs to be done. We can do it now or later. @mchacki WDYT?

File: arangod/Aql/KShortestPathsNode.cpp [hunk 2]
```diff
@@ -328,16 +337,60 @@ std::unique_ptr<ExecutionBlock> KShortestPathsNode::createBlock(
...
+            KShortestPathsExecutorInfos(outputRegister, engine.getQuery(),
+                                        std::move(kPathUnique), std::move(sourceInput),
+                                        std::move(targetInput));
+        return std::make_unique<ExecutionBlockImpl<KShortestPathsExecutor<KPathRefactored>>>(
+            &engine, this, std::move(registerInfos), std::move(executorInfos));
+      } else {
+        using KPathRefactored =
+            TwoSidedEnumerator<QueueTracer<FifoQueue<SingleServerProvider::Step>>,
+                               PathStoreTracer<PathStore<SingleServerProvider::Step>>, ProviderTracer<SingleServerProvider>>;
+        // TODO: below copy paste from above. clean this up later.
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:368]: Any idea to do this properly? Using a local method? Or a better way available?

File: arangod/Aql/KShortestPathsNode.cpp [hunk 3]
```diff
@@ -328,16 +337,60 @@ std::unique_ptr<ExecutionBlock> KShortestPathsNode::createBlock(
...
 
   if (shortestPathType() == arangodb::graph::ShortestPathType::Type::KPaths) {
-    auto finder = std::make_unique<graph::KPathFinder>(*opts);
-    auto executorInfos =
-        KShortestPathsExecutorInfos(outputRegister, std::move(finder),
-                                    std::move(sourceInput), std::move(targetInput));
-    return std::make_unique<ExecutionBlockImpl<KShortestPathsExecutor<graph::KPathFinder>>>(
-        &engine, this, std::move(registerInfos), std::move(executorInfos));
+    if (opts->refactor()) {
+      // Create IndexAccessor for BaseProviderOptions (TODO: Location need to be
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:341]: fyi open todo

File: arangod/Aql/KShortestPathsNode.cpp [hunk 4]
```diff
@@ -382,6 +436,82 @@ void KShortestPathsNode::kShortestPathsCloneHelper(ExecutionPlan& plan,
   c._toCondition = _toCondition->clone(_plan->getAst());
 }
 
+std::vector<arangodb::graph::IndexAccessor> KShortestPathsNode::buildUsedIndexes() const {
+  // std::vector<IndexAccessor> usedIndexes{
+  // IndexAccessor{edgeIndexHandle, indexCondition, 0}};
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:441]:
```suggestion
```

File: arangod/Aql/KShortestPathsNode.cpp [hunk 5]
```diff
@@ -382,6 +436,82 @@ void KShortestPathsNode::kShortestPathsCloneHelper(ExecutionPlan& plan,
...
+  std::vector<IndexAccessor> indexAccessors{};
+
+  if (_options->refactor()) {
+    // TODO: remove / cleanup later
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:445]: @mchacki fyi open todo.

File: arangod/Aql/KShortestPathsNode.cpp [hunk 6]
```diff
@@ -382,6 +436,82 @@ void KShortestPathsNode::kShortestPathsCloneHelper(ExecutionPlan& plan,
...
+          break;
+      }
+    }
+  }
+  return indexAccessors;
+}
+
+std::vector<arangodb::graph::IndexAccessor> KShortestPathsNode::buildReverseUsedIndexes() const {
+  // std::vector<IndexAccessor> usedIndexes{
+  // IndexAccessor{edgeIndexHandle, indexCondition, 0}};
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:479]:
```suggestion
```

File: arangod/Aql/KShortestPathsNode.cpp [hunk 7]
```diff
@@ -382,6 +436,82 @@ void KShortestPathsNode::kShortestPathsCloneHelper(ExecutionPlan& plan,
...
+  std::vector<IndexAccessor> indexAccessors{};
+
+  if (_options->refactor()) {
+    // TODO: remove / cleanup later
```
Comment (hkernbach) [inline arangod/Aql/KShortestPathsNode.cpp:483]: Open todo. Same as above.

File: arangod/Graph/Cache/RefactoredTraverserCache.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        << ".unload(); db." << col->name() << ".load()' in arangosh to fix this.";
+    TRI_ASSERT(false);  // for maintainer mode
+    return arangodb::velocypack::Slice::nullSlice();
+  }
+
+  return VPackSlice(_mmdr.vpack());
+}
+
+VPackSlice RefactoredTraverserCache::lookupVertexInCollection(arangodb::velocypack::HashedStringRef const& idHashed) {
+  // TODO: Missing produceVertices check
```
Comment (hkernbach) [inline arangod/Graph/Cache/RefactoredTraverserCache.cpp:90]: Some open TODOs here. Those needs to be checked before merge.

File: arangod/Graph/Cursors/RefactoredSingleServerEdgeCursor.cpp [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+          // _opts->cache()->increaseCounter(); TODO CHECK
+          callback(EdgeDocumentToken(cid, token), edgeDoc, cursorId);
+          return true;
+        });
+      });
+    }
+  }
+}
+
+// TODO maybe remove, unsure if we need it still.
```
Comment (hkernbach) [inline arangod/Graph/Cursors/RefactoredSingleServerEdgeCursor.cpp:239]: @mchacki fyi please check

File: arangod/Graph/Cursors/RefactoredSingleServerEdgeCursor.h [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+struct IndexAccessor const { //definition
+  IndexHande const& _index; // Muss das query vorher ausrechnen, der Provider muss es bekommen (pro collection)
+  AstNode* _searchCondition; // "" (pro index) _from == X  || _to == X
+}
+// keine ahnung von Rearm, baut den ersten Cursor?
+
+// Action.
+=> IndexCursor := _trx->indexScanForCondition()
+=> rearm
+*/
```
Comment (hkernbach) [inline arangod/Graph/Cursors/RefactoredSingleServerEdgeCursor.h:69]:
```suggestion
```

Comment (hkernbach) [inline arangod/Graph/Cursors/RefactoredSingleServerEdgeCursor.h:69]:
```suggestion
```

File: arangod/Graph/Providers/SingleServerProvider.cpp [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+      _cache(_trx.get(), &queryContext),
+      _opts(std::move(opts)) {
+  // activateCache(false); // TODO CHECK RefactoredTraverserCache
+  _cursor = buildCursor();
+}
+
+void SingleServerProvider::activateCache(bool enableDocumentCache) {
+  // Do not call this twice.
+  // TRI_ASSERT(_cache == nullptr);
+  // TODO: enableDocumentCache check + opts check + cacheManager check
```
Comment (hkernbach) [inline arangod/Graph/Providers/SingleServerProvider.cpp:95]: Open TODO: Need to check/verify cache.

File: arangod/Graph/Providers/SingleServerProvider.cpp [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+auto SingleServerProvider::startVertex(VertexType vertex) -> Step {
+  LOG_TOPIC("78156", TRACE, Logger::GRAPHS)
+      << "<MockGraphProvider> Start Vertex:" << vertex;
+  // Should not clear anything: clearCursor(vertex);
+
+  // Create default initial step
+
+  // TODO Implement minimal variant of _cache in the Provider
```
Comment (hkernbach) [inline arangod/Graph/Providers/SingleServerProvider.cpp:118]: TODO

File: arangod/Graph/Providers/SingleServerProvider.cpp [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+auto SingleServerProvider::fetch(std::vector<Step*> const& looseEnds)
+    -> futures::Future<std::vector<Step*>> {
+  LOG_TOPIC("78156", TRACE, Logger::GRAPHS)
+      << "<MockGraphProvider> Fetching...";
+  std::vector<Step*> result{};  // TODO: Question: Modify inplace or build up new result vector
+  result.reserve(looseEnds.size());
+  // TODO do we need to do something at all?
+  // We may want to cache the Vertex Data
```
Comment (hkernbach) [inline arangod/Graph/Providers/SingleServerProvider.cpp:130]: Check cache.

File: arangod/Graph/Providers/SingleServerProvider.h [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+ private:
+  // Unique_ptr to have this class movable, and to keep reference of trx()
+  // alive - Note: _trx must be first here because it is used in _cursor
+  std::unique_ptr<arangodb::transaction::Methods> _trx;
+
+  std::unique_ptr<RefactoredSingleServerEdgeCursor> _cursor;
+
+  arangodb::aql::QueryContext* _query;
+  // We DO take responsibility for the Cache (TODO?)
```
Comment (hkernbach) [inline arangod/Graph/Providers/SingleServerProvider.h:163]:
```suggestion
```

File: arangod/Graph/Queues/FifoQueue.h [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+#include <queue>
+
+namespace arangodb {
+namespace graph {
+
+template <class StepType>
+class FifoQueue {
+ public:
+  using Step = StepType;
+  // TODOS: Add Sorting (Performance)
```
Comment (hkernbach) [inline arangod/Graph/Queues/FifoQueue.h:38]: To be discussed.

**Review by hkernbach — COMMENTED** [review]
Still, a few TODOs left - needs to be discussed.

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13255/

**Comment by hkernbach** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13256/

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13348/

**Comment by hkernbach** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13357/

File: arangod/Transaction/Methods.h [hunk 16]
```diff
@@ -115,6 +116,9 @@ class Methods {
   /// @brief destroy the transaction
   virtual ~Methods();
 
+  Methods(Methods&&) = default;
```
Comment (mchacki) [inline arangod/Transaction/Methods.h:119]: Check if this is still needed,
I know we used it in between, but this may be obsolete now

**Review by mchacki — CHANGES_REQUESTED** [review]
Please see comments inline.
There are some leftovers.

File: arangod/Transaction/Methods.h (hunk 16 above)
Comment (mchacki) [reply arangod/Transaction/Methods.h:120]: It is needed

File: arangod/Aql/KShortestPathsNode.cpp (hunk 2 above)
Comment (mchacki) [reply arangod/Aql/KShortestPathsNode.cpp:372]: I was thinking of a Factory, but we can add this later if we have more algorithms

**Review by mchacki — APPROVED** [review]
LGTM.

As this is part of a framework there are still some glitches at not yet implemented parts.
But I consider this acceptable for now, as the relevant code is working as expected and good.

**Comment by mchacki** [comment]
Jenkins:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/13567/
All Blue

**Pull request merged**